verifier/verifier.go

```diff
@@ -662,51 +678,210 @@ func verifyExpiry(outcome *notation.VerificationOutcome) *notation.ValidationRes
+			timestampLowerLimit := time.Now()
+			timestampUpperLimit := timestampLowerLimit
+			for _, cert := range signerInfo.CertificateChain {
+				if timestampLowerLimit.Before(cert.NotBefore) {
+					return &notation.ValidationResult{
+						Error:  fmt.Errorf("verification time is before certificate %q validity period, it will be valid from %q", cert.Subject, cert.NotBefore.Format(time.RFC1123Z)),
+						Type:   trustpolicy.TypeAuthenticTimestamp,
+						Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeAuthenticTimestamp],
+					}
+				}
+				if timestampUpperLimit.After(cert.NotAfter) {
+					return &notation.ValidationResult{
+						Error:  fmt.Errorf("verification time is after certificate %q validity period, it was expired at %q", cert.Subject, cert.NotAfter.Format(time.RFC1123Z)),
+						Type:   trustpolicy.TypeAuthenticTimestamp,
+						Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeAuthenticTimestamp],
+					}
+				}
+			}
+			// success
+			return &notation.ValidationResult{
+				Type:   trustpolicy.TypeAuthenticTimestamp,
+				Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeAuthenticTimestamp],
+			}
+		}
+		// Performing timestamp verification
+		// 1. Timestamp countersignature MUST be present
+		logger.Info("Checking timestamp countersignature existence...")
+		if len(signerInfo.UnsignedAttributes.TimestampSignature) == 0 {
+			return &notation.ValidationResult{
+				Error:  errors.New("no timestamp countersignature was found in the signature envelope"),
+				Type:   trustpolicy.TypeAuthenticTimestamp,
+				Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeAuthenticTimestamp],
+			}
+		}
+		// 2. Verify the timestamp countersignature
+		logger.Info("Verifying the timestamp countersignature...")
+		signedToken, err := tspclient.ParseSignedToken(signerInfo.UnsignedAttributes.TimestampSignature)
+		if err != nil {
+			return &notation.ValidationResult{
+				Error:  fmt.Errorf("failed to parse timestamp countersignature with error: %w", err),
+				Type:   trustpolicy.TypeAuthenticTimestamp,
+				Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeAuthenticTimestamp],
+			}
+		}
+		info, err := signedToken.Info()
+		if err != nil {
+			return &notation.ValidationResult{
+				Error:  fmt.Errorf("failed to get the timestamp TSTInfo with error: %w", err),
+				Type:   trustpolicy.TypeAuthenticTimestamp,
+				Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeAuthenticTimestamp],
+			}
+		}
+		trustTSACerts, err := loadX509TSATrustStores(ctx, outcome.EnvelopeContent.SignerInfo.SignedAttributes.SigningScheme, policyName, trustStores, x509TrustStore)
+		if err != nil {
+			return &notation.ValidationResult{
+				Error:  fmt.Errorf("failed to load tsa trust store with error: %w", err),
+				Type:   trustpolicy.TypeAuthenticTimestamp,
+				Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeAuthenticTimestamp],
+			}
+		}
+		if len(trustTSACerts) < 1 {
```

Comment:
nit: reduce mental strain for reviewers
```suggestion
		if len(trustTSACerts) == 0 {
```